mitiq/cdr/test_clifford_training_data.py

```diff
@@ -0,0 +1,254 @@
+        _select([], fraction_non_clifford=0.0, method="unknown method")
+
+
+@pytest.mark.parametrize("method", ("closest", "uniform", "gaussian"))
+def test_replace(method):
+    q = cirq.LineQubit(0)
+    ops = [cirq.ops.rz(0.01).on(q), cirq.ops.rz(-0.77).on(q)]
+
+    new_ops = _replace(
+        non_clifford_ops=ops,
+        method=method,
+        random_state=np.random.RandomState(1),
+    )
+
+    assert len(new_ops) == len(ops)
+    assert all(_is_clifford(op) for op in new_ops)
+
+
+def test_map_to_near_clifford():
+    q = cirq.LineQubit(0)
+    ops = [cirq.ops.rz(np.pi / 2.0 + 0.01).on(q), cirq.ops.rz(-0.22).on(q)]
+
+    new_ops = _map_to_near_clifford(
+        ops,
+        fraction_non_clifford=0.0,
+        method_select="uniform",
+        method_replace="uniform",
+        seed=1,
+    )
+    expected_ops = [cirq.rz(np.pi * 0.5).on(q), cirq.rz(0.0).on(q)]
+    assert new_ops == expected_ops
+
+
+def test_generate_training_circuits_bad_methods():
+    with pytest.raises(ValueError):
+        generate_training_circuits(
+            Circuit(cirq.ops.rx(0.5).on(cirq.LineQubit(0))),
+            num_training_circuits=1,
+            fraction_non_clifford=0.0,
+            method_select="unknown select method",
+        )
+
+    with pytest.raises(ValueError):
+        generate_training_circuits(
+            Circuit(cirq.ops.rx(0.5).on(cirq.LineQubit(0))),
+            num_training_circuits=1,
+            fraction_non_clifford=0.0,
+            method_replace="unknown replace method",
+        )
+
+
+def test_generate_training_circuits_with_clifford_circuit():
+    with pytest.raises(ValueError, match="Circuit is already Clifford."):
+        generate_training_circuits(
+            Circuit(cirq.ops.rx(0.0).on(cirq.LineQubit(0))),
+            num_training_circuits=1,
+            fraction_non_clifford=0.0,
+        )
+
+
+@pytest.mark.parametrize("method_select", ["uniform", "gaussian"])
+@pytest.mark.parametrize("method_replace", ["uniform", "gaussian", "closest"])
+@pytest.mark.parametrize(
+    "kwargs", [{}, {"sigma_select": 0.5, "sigma_replace": 0.5}]
+)
+def test_generate_training_circuits_mega(
+    method_select, method_replace, kwargs
+):
+    circuit = random_x_z_cnot_circuit(qubits=4, n_moments=10, random_state=1)
+    num_train = 10
+    fraction_non_clifford = 0.1
+
+    train_circuits = generate_training_circuits(
+        circuit,
+        num_training_circuits=num_train,
+        fraction_non_clifford=0.1,
+        random_state=np.random.RandomState(13),
+        method_select=method_select,
+        method_replace=method_replace,
+        **kwargs,
+    )
+    assert len(train_circuits) == num_train
+
+    for train_circuit in train_circuits:
+        assert set(train_circuit.all_qubits()) == set(circuit.all_qubits())
+        assert count_non_cliffords(train_circuit) == int(
+            fraction_non_clifford * count_non_cliffords(circuit)
+        )
```

Comment:
```suggestion
        assert count_non_cliffords(train_circuit) == int(
            round(fraction_non_clifford * count_non_cliffords(circuit))
        )
```
If round is used in the code of `generate_training_circuits`, this may be necessary.